src/chipper/keyboard2.cljs

```diff
@@ -0,0 +1,233 @@
+;; "something" is probably just a vec? or a map?
+;; like [property parameters]
+;; property and parameters would be like
+;; :position  [dline dchan dattr]  ; let's say numerical values + :first :last
+;; :attribute [line chan attr val] ; attr is any of the three note position
+;; :mode      :edit                ; :normal :edit :visual
+;; (and it's called edit because it behaves like replace, which doesn't make sense
+;; as a name in the absence of a distinct insert mode)
+;; :special ; which is then its own thing, like {:action :octave-up :parameters 10}
+;; :custom  ; or maybe `:macro`? something to facilitate more complex things
+;; then a big handler decides which little handler actually alters state
+;; big handler would basically just be a another map like
+;; {:position  set-position!
+;;  :attribute set-attribute!  ; this needs a better name than attribute, maybe
+;;  etc}
+
+;; Available properties are
+;; :position :frame :mode :attr :macro :special
+;; Available macro types are
+;; :simple   (a sequence of position/frame/mode/attr)
+;; :advanced (a callable?)
+;; and macros mappings are defined as
+;; {:Trigger [:macro {:type type :value value}]}
+;; See edit mode :Backspace for an example
+
+(def measure-size 4)
+
+;; XXX These should maybe be somwhere else, like core, or... something
+;; also note that if you use these, the max index of the cursor is 1 less
+;; because position is 0-indexed
+(def frame-count 32)
+(def frame-length 32)
+(def chan-count 4)
+(def attr-count 3)
+;; a hack, because of course
+(def total-attr-positions (* chan-count attr-count))
+
+(def -motions
+  {:down  [ 1  0  0]
+   :up    [-1  0  0]
+   :left  [ 0  0 -1]
+   :right [ 0  0  1]})
+
+(def normal-keymap
+  {;; relative movement
+   ;; keeping :position for clarity. The representation for cursor position may
+   ;; change in the future, thus -motions
+   :KeyJ        [:relative-position (:down  -motions)]
+   :KeyK        [:relative-position (:up    -motions)]
+   :KeyH        [:relative-position (:left  -motions)]
+   :KeyL        [:relative-position (:right -motions)]
+
+   :KeyEnter    [:relative-position (:down  -motions)]
+   :ArrowDown   [:relative-position (:down  -motions)]
+   :ArrowUp     [:relative-position (:up    -motions)]
+   :ArrowLeft   [:relative-position (:left  -motions)]
+   :ArrowRight  [:relative-position (:right -motions)]
+
+   :KeyW        [:relative-position [measure-size     0 0]]
+   :KeyB        [:relative-position [(- measure-size) 0 0]]
+
+   :ShiftBracketRight [:frame  1]
+   :ShiftBracketLeft  [:frame -1]
+
+   ;; absolute movement
+   :KeyG        [:absolute-position [:first :first :first]]
+   :ShiftKeyG   [:absolute-position [:last  :last  :last]]
+   :Digit0      [:absolute-position [:first :first :first]]
+   :ShiftDigit4 [:absolute-position [:last  :last  :last]]
+
+   ;; mode change
+   :KeyI [:mode :edit]
+   :KeyV [:mode :visual]
```

Comment:
shift-v? i assume the only meaningful modes are v-line and v-block, so like v-frame and v-chan

Reply:
yeah, should be shiftv. also missing c-v